drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,431 @@
+	/* I2C error has occurred */
+	I2C_CC32XX_ERROR = 0xFF
+};
+
+struct i2c_cc32xx_config {
+	u32_t base;
+	u32_t bitrate;
+};
+
+struct i2c_cc32xx_data {
+	struct k_sem mutex;
+	struct k_sem transfer_complete;
+
+	volatile enum i2c_cc32xx_state state;
+
+	u8_t  *write_buf_idx;  /* Internal inc. writeBuf index */
+	size_t write_cnt;      /* Internal dec. writeCounter */
+	u8_t  *read_buf_idx;   /* Internal inc. readBuf index */
+	size_t read_cnt;       /* Internal dec. readCounter */
+	u16_t  slave_addr;     /* Cache slave address for ISR use */
+};
+
+static void configure_i2c_irq(void);
+
+static int i2c_cc32xx_configure(struct device *dev, u32_t dev_config_raw)
+{
+	u32_t base = DEV_BASE(dev);
+	union dev_config dev_config = (union dev_config)dev_config_raw;
+	u32_t bitrate_id;
+
+	if (!dev_config.bits.is_master_device) {
+		return -EINVAL;
+	}
+
+	if (dev_config.bits.use_10_bit_addr) {
+		return -EINVAL;
+	}
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		bitrate_id = 0;
+		break;
+	case I2C_SPEED_FAST:
+		bitrate_id = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	MAP_I2CMasterInitExpClk(base, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+				bitrate_id);
+
+	return 0;
+}
+
+static void _i2c_cc32xx_prime_transfer(struct device *dev, struct i2c_msg *msg,
+				      u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t base = DEV_BASE(dev);
+
+	/* Initialize internal counters and buf pointers: */
+	if (msg->flags & I2C_MSG_READ) {
+		data->read_buf_idx = msg->buf;
+		data->read_cnt = msg->len;
+	} else {
+		data->write_buf_idx = msg->buf;
+		data->write_cnt = msg->len;
+	}
+	data->slave_addr = addr;
+
+	/* Start transfer in Transmit mode */
+	if (data->write_cnt) {
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, false);
+
+		/* Update the I2C state */
+		data->state = I2C_CC32XX_WRITE_MODE;
+
+		/* Write data contents into data register */
+		MAP_I2CMasterDataPut(base, *((data->write_buf_idx)++));
+
+		/* Start the I2C transfer in master transmit mode */
+		MAP_I2CMasterControl(base, I2C_MASTER_CMD_BURST_SEND_START);
+
+	} else {
+		/* Start transfer in Receive mode */
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, true);
+
+		/* Update the I2C mode */
+		data->state = I2C_CC32XX_READ_MODE;
+
+		if (data->read_cnt < 2) {
+			/* Start the I2C transfer in master receive mode */
+			MAP_I2CMasterControl(base,
+				       I2C_MASTER_CMD_BURST_RECEIVE_START_NACK);
+		} else {
+			/* Start the I2C transfer in burst receive mode */
+			MAP_I2CMasterControl(base,
+					    I2C_MASTER_CMD_BURST_RECEIVE_START);
+		}
+	}
+}
+
+static int i2c_cc32xx_transfer(struct device *dev, struct i2c_msg *msgs,
+			       u8_t num_msgs, u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	int retval = 0;
+
+	/* Iterate over all the messages */
+	for (int i = 0; i < num_msgs; i++) {
+		/* Acquire the driver mutex */
+		k_sem_take(&data->mutex, K_FOREVER);
```

Comment:
I believe the mutex should be outside of the loop, otherwise you allow a combined I2C write/read transfer to be interrupted by some unrelated message.

Reply:
Good point.  The temp sensor driver typically writes a register location first to read, immediately followed by a read op.  Probably a good idea to not interrupt that.